Bit-Quill/amazon-timestream-tools
# Return 2.0 formatted Lambda responses for local invocation

- The environment variable `local_invocation` has been added and is set to `true` by default when the Lambda function is run with `cargo lambda watch`. This environment enables responses to be returned in a format `cargo lambda` expects, in the 2.0 formatting. Otherwise, the 1.0 format will be returned, which the synchronous API Gateway expects.
- `cargo fmt` run.

- [x] Tested with synchronous invocation.
- [x] Tested with asynchronous invocation.

## Test Steps

The goal of this PR is to return the expected Lambda responses for different services.

With the Lambda proxy integration used for synchronous invocation, the REST API Gateway will expect a 1.0 Lambda response, a response without a `"cookie"` field. The REST API Gateway will return an error if this field is present.

When run locally, `cargo lambda` expects the Lambda function to return a 2.0 Lambda response, a response including a `"cookie"` field. `cargo lambda` will return an error if this field is not present.

Ingestion should succeed when run locally, with synchronous deployment, and asynchronous deployment.

### Local Deployment

1. **IMPORTANT**. Change the value of `region` in `Cargo.toml` to `"us-west-2"`. This is crucial to not interfere with ongoing product testing.
2. Run the following command to run the connector locally, making sure that `us-west-2` is output as the region the client connects to:
    ```shell
    cargo lambda watch
    ```
3. In `../sample-influxdb-clients/go/` run:
    ```shell
    go run line-protocol-client-demo.go --region us-west-2
    ```
4. The client will output an error if one occurs.

### SAM Synchronous Deployment

1. Run the following command to package the connector:
    ```shell
    cargo lambda build --output-format zip --arm64 --release
    ```
3. Run the following command to deploy the connector with synchronous invocation enabled:
    ```shell
    sam deploy template.yml \
        --region us-west-2 \
        --stack-name InfluxDBTimestreamConnectorTest \
        --resolve-s3 \
        --capabilities CAPABILITY_IAM \
        --parameter-overrides \
            EnableAsyncInvocation=false
    ```
4. Take note of the `Endpoint` output.
5. In `../sample-influxdb-clients/go/` run:
    ```shell
    go run line-protocol-client-demo.go \
        --region us-west-2 \
        --endpoint <Endpoint value> \
        --service execute-api
    ```
6. The client will output an error if one occurs.
7. Delete the stack:
    ```shell
    sam delete \
        --stack-name InfluxDBTimestreamConnectorTest \
        --region us-west-2
    ```

### SAM Asynchronous Deployment

1. Run the following command to package the connector:
    ```shell
    cargo lambda build --output-format zip --arm64 --release
    ```
2. Run the following command to deploy the connector with asynchronous invocation enabled:
    ```shell
    sam deploy template.yml \
        --region us-west-2 \
        --stack-name InfluxDBTimestreamConnectorTest \
        --resolve-s3 \
        --capabilities CAPABILITY_IAM
    ```
3. Take note of the `Endpoint` and `LambdaDeadLetterQueueName` outputs.
4. In `../sample-influxdb-clients/go/` run:
    ```shell
    go run line-protocol-client-demo.go \
        --region us-west-2 \
        --endpoint <Endpoint value> \
        --service execute-api
    ```
5. The client will not output an error if one occurs. Instead, in the AWS console navigate to the SQS queue with the same name as the `LambdaDeadLetterQueueName` and click "Send and receive messages" and then "Poll for messages" to view any errors.
6. Delete the stack:
    ```shell
    sam delete \
        --stack-name InfluxDBTimestreamConnectorTest \
        --region us-west-2
    ```
